### DISCUSSION TITLE
Repository type for arbitrary file, e.g., a PHAR?

### DISCUSSION THREAD

**Discussion by TravisCarden** [composer/composer#9976] [Q&A / Support] [answered]
tl;dr: I want to require a remote PHAR file as a dependency and install it intact, without decompressing it. Is there a way to do this, such as with a custom repository?

I want to use https://github.com/phpDocumentor/phpDocumentor, but installing it as a Composer requirement is neither recommended nor supported due to its own huge dependency list. I suggested a shim in https://github.com/phpDocumentor/phpDocumentor/issues/2930, but in the meantime I'm wondering if there's a way I can fake one myself, such as with a custom repository.  However, I haven't found a way. I was kind of hoping there was a "phar" repository type, but alas I don't see such a thing documented anywhere. "artifact" sounded promising but didn't turn out to be what I was looking for. I tried the below (in case "magic"), but it treats the PHAR as a ZIP and decompresses it.

```json
    "repositories": [
        {
            "type": "package",
            "package": {
                "name": "my/phpdocumentor-shim",
                "version": "1.0.0",
                "dist": {
                    "url": "https://phpdoc.org/phpDocumentor.phar",
                    "type": "phar"
                }
            }
        }
    ],
```

Right now I'm doing this:

```json
"scripts": {
    "post-update-cmd": "@install-phpdocumentor",
    "install-phpdocumentor": "php -r \"copy('https://phpdoc.org/phpDocumentor.phar', 'vendor/bin/phpdocumentor.phar');\""
}
```

Is there a better way to accomplish this? If not, I'll create an issue to propose a new type of "file" or something.

**Comment by Seldaek** [comment, marked as answer]
There is a type "file" already, which should simply put the phar file where you need it.

e.g.

```
{
    "require": {
        "my/phpdocumentor-shim": "*"
    },
    "repositories": [
        {
            "type": "package",
            "package": {
                "name": "my/phpdocumentor-shim",
                "version": "1.0.0",
                "dist": {
                    "url": "https://phpdoc.org/phpDocumentor.phar",
                    "type": "file"
                },
                "bin": ["phpDocumentor.phar"]
            }
        }
    ]
}
```

```
composer update
vendor/bin/phpDocumentor.phar
```

The only downside is you can't control the output filename, it takes it from the filename in the URL, so you're stuck with the capitalized name.

**Reply by TravisCarden** [reply]
Ha! OK thanks. 😄

In that case, I'll just point out that the "file" repository type is not documented at https://getcomposer.org/doc/05-repositories.md.

Thank you.

**Reply by stof** [reply]
this is not a repository type. It is a dist type

**Reply by stof** [reply]
However, `dist` and `source` are not documented on https://getcomposer.org/doc/04-schema.md
